src/components/MessageList/LivestreamMessageList.js

```diff
@@ -0,0 +1,502 @@
+    const scrollToBottom = hasNewMessage && (isOwner || !userScrolledUp);
+
+    if (scrollToBottom) {
+      if (this.props.messages.length > this.props.messageLimit) {
+        this.setState({
+          slicedMessages: this.props.messages.slice(-this.props.messageLimit),
+        });
+      } else {
+        this.setState({
+          slicedMessages: this.props.messages,
+        });
+      }
+
+      this.scrollToBottom();
+
+      // remove the scroll notification if we already scrolled down...
+      if (this.state.newMessagesNotification)
+        this.setState({ newMessagesNotification: false });
+
+      return;
+    }
+
+    // Making following state updates as async, so that when we update scroll position,
+    // we have the updated scrollHeight of the list.
+    if (hasNewMessage) {
+      await this.setAsyncState({
+        slicedMessages: this.props.messages.slice(
+          -(this.state.slicedMessages.length + 1),
+        ),
+      });
+    } else if (prevProps.messages.length < this.props.messages.length) {
+      await this.setAsyncState({
+        slicedMessages: this.props.messages,
+      });
+    }
+
+    if (snapshot !== null) {
+      // Maintain the offsetTop of scroll so that content in viewport doesn't move.
+      // This is for the case where user has scroll up significantly and a new message arrives from someone.
+      if (hasNewMessage) {
+        this.scrollToTarget(snapshot.offsetTop, this.messageList.current);
+      } else {
+        // Maintain the bottomOffset of scroll.
+        // This is for the case of pagination, when more messages get loaded.
+        this.scrollToTarget(
+          list.scrollHeight - snapshot.offsetBottom,
+          this.messageList.current,
+        );
+      }
+    }
+
+    // Check the scroll position... if you're scrolled up show a little notification
+    if (hasNewMessage && !this.state.newMessagesNotification) {
+      this.setState({ newMessagesNotification: true });
+    }
+  }
+
+  setAsyncState = (state) => {
+    return new Promise((resolve) => {
+      this.setState(state, resolve);
+    });
+  };
+
+  scrollToBottom = () => {
+    this._scrollToRef(this.bottomRef, this.messageList);
+  };
+
+  _scrollToRef = (el, parent) => {
+    const scrollDown = () => {
+      if (el && el.current && parent && parent.current) {
+        this.scrollToTarget(el.current, parent.current);
+      }
+    };
+
+    scrollDown();
+    // scroll down after images load again
+    setTimeout(scrollDown, 200);
+  };
+
+  /**
+   * target - target to scroll to (DOM element, scrollTop Number, 'top', or 'bottom'
+   * containerEl - DOM element for the container with scrollbars
+   * source: https://stackoverflow.com/a/48429314
+   */
+  scrollToTarget = (target, containerEl) => {
+    // Moved up here for readability:
+    const isElement = target && target.nodeType === 1;
+    const isNumber =
+      Object.prototype.toString.call(target) === '[object Number]';
```

Comment:
Using `toString()` to check types is unreliable (the `toStringTag` can be changed to anything). How about using `typeof` instead?